File: android/app/src/main/java/com/david/tehilim/core/service/PrayerShareLink.kt
package com.david.tehilim.core.service

import android.content.Context
import android.net.Uri
import com.david.tehilim.R
import com.david.tehilim.core.model.PrayerType
import com.david.tehilim.core.model.RelationType
import com.david.tehilim.core.model.SavedPrayerIntent
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Mirror Android du PrayerShareLink.swift iOS.
 *
 * Encode / décode une prière partageable via lien `tehilim://prayer`.
 * Le lien est **minimal** : type, prénom du proche, lien de parenté, prénom
 * de la mère, date du décès optionnelle. La séquence de lettres et la date
 * hébraïque sont **recalculées à l'import** (elles dérivent de ces champs).
 *
 * Format : `tehilim://prayer?v=1&type=<malade|defunt>&name=<enc>&rel=<ben|bat>&mother=<enc>&death=<yyyy-MM-dd>`
 *
 * Le schéma `tehilim://` est déjà déclaré dans l'AndroidManifest (intent-filter
 * `<data android:scheme="tehilim" />`, sans host → capture tous les hosts).
 */
object PrayerShareLink {

    const val SCHEME = "tehilim"
    const val HOST = "prayer"
    const val VERSION = "1"

    /** Page de redirection https (GitHub Pages) — un lien https est cliquable
     *  dans Mail / WhatsApp / SMS (contrairement à tehilim://). La page rouvre
     *  ensuite tehilim://prayer?... avec les mêmes paramètres. */
    const val WEB_BASE_URL = "https://dadoou5.github.io/p/"

    /** `yyyy-MM-dd` / locale US : format stable, identique à iOS. */
    private fun dateFormat() = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    // MARK: - Encodage

    fun uri(intent: SavedPrayerIntent): Uri {
        // Lien de partage = page de redirection https (cliquable partout).
        val builder = Uri.parse(WEB_BASE_URL).buildUpon()
            .appendQueryParameter("v", VERSION)
            .appendQueryParameter("type", intent.prayerType.wire)
            .appendQueryParameter("name", intent.relativeFirstName)
            .appendQueryParameter("rel", intent.relationType.wire)
            .appendQueryParameter("mother", intent.motherFirstName)
        intent.civilDateOfDeathEpochMillis?.let {
            builder.appendQueryParameter("death", dateFormat().format(Date(it)))
        }
        return builder.build()
    }

    /** Message texte prêt à partager (SMS / WhatsApp) : description, date du
     *  décès + prochaine azcara (langue de l'app), puis le lien d'import. */
    fun shareMessage(context: Context, intent: SavedPrayerIntent): String {
        val link = uri(intent).toString()
        val intro = context.getString(R.string.share_prayer_intro)
        val df = java.text.DateFormat.getDateInstance(java.text.DateFormat.LONG, Locale.getDefault())
        val sb = StringBuilder()
        sb.append("${intent.prayerType.saveActionTitle} — ${intent.hebrewSubject}")
        intent.civilDateOfDeathEpochMillis?.let { millis ->
            sb.append("\n").append(context.getString(R.string.share_date_of_death, df.format(Date(millis))))
            MemorialCalculator.nextYahrzeit(Date(millis))?.let { next ->
                sb.append("\n").append(context.getString(R.string.share_next_azcara, df.format(next)))
            }
        }
        sb.append("\n\n").append(intro).append("\n").append(link)
        return sb.toString()
    }

    // MARK: - Décodage

    data class Payload(
        val prayerType: PrayerType,
        val relativeFirstName: String,
        val relationType: RelationType,
        val motherFirstName: String,
        val civilDateOfDeathEpochMillis: Long?
    ) {
        val hebrewSubject: String
            get() = "$relativeFirstName ${relationType.hebrew} $motherFirstName"
    }

    /** True si l'URI est un lien de prière géré : schéma custom
     *  `tehilim://prayer` OU App Link `https://…/p/…`. */
    fun isPrayerLink(uri: Uri): Boolean {
        if (uri.scheme == SCHEME && uri.host == HOST) return true
        // Accepte « /p » exact ET « /p/… » (selon que le « / » final est
        // conservé ou non), sans matcher « /privacy ».
        val p = uri.path
        return uri.scheme == "https" && (p == "/p" || p?.startsWith("/p/") == true)
    }

    /** Parse un lien de prière (`tehilim://prayer?...` ou `https://…/p/?...`). */
    fun payload(uri: Uri): Payload? {
        if (!isPrayerLink(uri)) return null
        val type = uri.getQueryParameter("type")?.let { prayerTypeFromWire(it) } ?: return null
        val rel = uri.getQueryParameter("rel")?.let { relationFromWire(it) } ?: return null
        val name = uri.getQueryParameter("name")?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        val mother = uri.getQueryParameter("mother")?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        val death = uri.getQueryParameter("death")?.let {
            runCatching { dateFormat().parse(it)?.time }.getOrNull()
        }
        return Payload(type, name, rel, mother, death)
    }

    /**
     * Construit un [SavedPrayerIntent] complet : régénère la séquence et la
     * date hébraïque. Rappels désactivés par défaut (le destinataire décide).
     */
    fun makeIntent(p: Payload): SavedPrayerIntent {
        val sequence = LetterSequenceGenerator.generate(
            relativeName = p.relativeFirstName,
            relation = p.relationType,
            motherName = p.motherFirstName,
            prayerType = p.prayerType
        )
        return SavedPrayerIntent(
            title = LetterSequenceGenerator.makeTitle(
                prayerType = p.prayerType,
                relativeName = p.relativeFirstName,
                relation = p.relationType,
                motherName = p.motherFirstName
            ),
            prayerType = p.prayerType,
            relativeFirstName = p.relativeFirstName,
            relationType = p.relationType,
            motherFirstName = p.motherFirstName,
            generatedLetters = sequence,
            civilDateOfDeathEpochMillis = p.civilDateOfDeathEpochMillis,
            hebrewDateOfDeath = p.civilDateOfDeathEpochMillis?.let {
                MemorialCalculator.hebrewYMD(Date(it))
            },
            remindersEnabled = false
        )
    }

    // MARK: - Valeurs « wire » (identiques à iOS rawValue)

    private val PrayerType.wire: String
        get() = when (this) {
            PrayerType.MALADE -> "malade"
            PrayerType.DEFUNT -> "defunt"
        }

    private val RelationType.wire: String
        get() = when (this) {
            RelationType.BEN -> "ben"
            RelationType.BAT -> "bat"
        }

    private fun prayerTypeFromWire(s: String): PrayerType? = when (s.lowercase()) {
        "malade" -> PrayerType.MALADE
        "defunt" -> PrayerType.DEFUNT
        else -> null
    }

    private fun relationFromWire(s: String): RelationType? = when (s.lowercase()) {
        "ben" -> RelationType.BEN
        "bat" -> RelationType.BAT
        else -> null
    }
}

File: ios/Tehilim/Core/Services/PrayerShareLink.swift
import Foundation

/// Encodage / décodage d'une prière partageable via lien `tehilim://prayer`.
///
/// Le lien est volontairement **minimal** : seuls les champs « source » sont
/// transmis (type, prénom du proche, lien de parenté, prénom de la mère,
/// date du décès optionnelle). La séquence de lettres (`generatedLetters`)
/// et la date hébraïque (`hebrewDateOfDeath`) sont **recalculées à l'import**
/// — elles dérivent déterministiquement de ces champs, inutile de les
/// transporter (URL plus courte, robuste aux évolutions du générateur).
///
/// Format : `tehilim://prayer?v=1&type=<malade|defunt>&name=<enc>&rel=<ben|bat>&mother=<enc>&death=<yyyy-MM-dd>`
///
/// Le schéma `tehilim://` est déjà enregistré (CFBundleURLSchemes) et géré
/// par `RootTabView.handleDeepLink`.
enum PrayerShareLink {

    static let scheme = "tehilim"
    static let host = "prayer"
    /// Version du format de payload — permet d'évoluer sans casser les liens
    /// déjà partagés.
    static let version = "1"

    /// Page de redirection https (GitHub Pages) — un lien `https://` est
    /// cliquable dans Mail / WhatsApp / SMS (contrairement à `tehilim://`).
    /// La page lit les mêmes paramètres et rouvre `tehilim://prayer?...`.
    static let webBaseURL = "https://tehilimapp.com/p/"

    /// `yyyy-MM-dd` en calendrier grégorien + locale POSIX : format stable,
    /// identique côté Android, indépendant de la locale de l'appareil.
    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.calendar = Calendar(identifier: .gregorian)
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = .current
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    // MARK: - Encodage

    /// Construit le lien de partage `https://…/p/?...` (page de redirection)
    /// représentant l'intent. Cliquable dans tous les messageries.
    static func url(for intent: SavedPrayerIntent) -> URL? {
        guard var comps = URLComponents(string: webBaseURL) else { return nil }
        var items: [URLQueryItem] = [
            URLQueryItem(name: "v", value: version),
            URLQueryItem(name: "type", value: intent.prayerType.rawValue),
            URLQueryItem(name: "name", value: intent.relativeFirstName),
            URLQueryItem(name: "rel", value: intent.relationType.rawValue),
            URLQueryItem(name: "mother", value: intent.motherFirstName)
        ]
        if let death = intent.civilDateOfDeath {
            items.append(URLQueryItem(name: "death", value: dateFormatter.string(from: death)))
        }
        comps.queryItems = items
        return comps.url
    }

    /// Message texte prêt à partager (SMS / WhatsApp) : description lisible,
    /// date du décès + prochaine azcara (dans la langue de l'expéditeur),
    /// puis le lien d'import. Taper le lien ouvre l'app et propose l'import.
    static func shareMessage(for intent: SavedPrayerIntent) -> String {
        let fr = AppLocale.code == "fr"   // hébreu → message EN
        let link = url(for: intent)?.absoluteString ?? ""
        let df = DateFormatter()
        df.locale = AppLocale.locale
        df.dateStyle = .long

        var lines: [String] = ["\(intent.prayerType.saveActionTitle) — \(intent.hebrewSubject)"]
        if let death = intent.civilDateOfDeath {
            lines.append((fr ? "Date de décès : " : "Date of death: ") + df.string(from: death))
            if let next = MemorialCalculator.nextYahrzeit(deathCivil: death) {
                lines.append((fr ? "Prochaine azcara : " : "Next azcara: ") + df.string(from: next)
                             + (fr ? " (commence la veille au soir)" : " (begins the previous evening)"))
            }
        }
        lines.append("")
        lines.append(fr ? "Ouvre cette prière dans Tehilim :" : "Open this prayer in Tehilim:")
        lines.append(link)
        return lines.joined(separator: "\n")
    }

    // MARK: - Décodage

    /// Payload décodé depuis un lien partagé.
    /// `Identifiable` pour piloter une présentation `.sheet(item:)`.
    struct Payload: Identifiable {
        let id = UUID()
        let prayerType: PrayerType
        let relativeFirstName: String
        let relationType: RelationType
        let motherFirstName: String
        let civilDateOfDeath: Date?

        var hebrewSubject: String {
            "\(relativeFirstName) \(relationType.hebrew) \(motherFirstName)"
        }
    }

    /// True si l'URL est un lien de prière géré : schéma custom
    /// `tehilim://prayer` OU Universal Link `https://…/p/…`.
    static func isPrayerLink(_ url: URL) -> Bool {
        if url.scheme == scheme && url.host == host { return true }
        // `URL.path` retire le « / » final : `https://…/p/?...` donne « /p ».
        // On accepte donc « /p » exact ET « /p/… », sans matcher « /privacy ».
        if url.scheme == "https" && (url.path == "/p" || url.path.hasPrefix("/p/")) { return true }
        return false
    }

    /// Parse un lien de prière (`tehilim://prayer?...` ou `https://…/p/?...`).
    /// Retourne nil si les champs requis sont manquants ou invalides.
    static func payload(from url: URL) -> Payload? {
        guard isPrayerLink(url),
              let comps = URLComponents(url: url, resolvingAgainstBaseURL: false)
        else { return nil }

        let q = Dictionary(
            (comps.queryItems ?? []).map { ($0.name, $0.value ?? "") },
            uniquingKeysWith: { first, _ in first }
        )

        guard let typeRaw = q["type"], let type = PrayerType(rawValue: typeRaw),
              let relRaw = q["rel"], let rel = RelationType(rawValue: relRaw),
              let name = q["name"]?.trimmingCharacters(in: .whitespacesAndNewlines), !name.isEmpty,
              let mother = q["mother"]?.trimmingCharacters(in: .whitespacesAndNewlines), !mother.isEmpty
        else { return nil }

        let death = q["death"].flatMap { dateFormatter.date(from: $0) }
        return Payload(
            prayerType: type,
            relativeFirstName: name,
            relationType: rel,
            motherFirstName: mother,
            civilDateOfDeath: death
        )
    }

    /// Construit un `SavedPrayerIntent` complet à partir d'un payload :
    /// régénère la séquence de lettres et la date hébraïque. Les rappels sont
    /// désactivés par défaut — le destinataire choisit lui-même de les activer.
    static func makeIntent(from p: Payload) -> SavedPrayerIntent {
        let sequence = LetterSequenceGenerator.generate(
            relativeName: p.relativeFirstName,
            relation: p.relationType,
            motherName: p.motherFirstName,
            prayerType: p.prayerType
        )
        return SavedPrayerIntent(
            title: LetterSequenceGenerator.makeTitle(
                prayerType: p.prayerType,
                relativeName: p.relativeFirstName,
                relation: p.relationType,
                motherName: p.motherFirstName
            ),
            prayerType: p.prayerType,
            relativeFirstName: p.relativeFirstName,
            relationType: p.relationType,
            motherFirstName: p.motherFirstName,
            generatedLetters: sequence,
            civilDateOfDeath: p.civilDateOfDeath,
            hebrewDateOfDeath: p.civilDateOfDeath.map { MemorialCalculator.hebrewYMD(from: $0) },
            remindersEnabled: false,
            notifySevenDaysBefore: true,
            notifySameDay: true
        )
    }
}
